File: LeetCode/medium/Generate_Parentheses_recursive.cxx
class Solution {
    vector<string> res;
public:
    void solve(string s, int n, int a, int b)
    {
        if (s.size() == n*2)
        {
            res.push_back(s);
        }
        else
        {
            if (a < n)
            {
                solve(s+"(", n, a+1, b);
            }
            if (a > b)
            {
                solve(s+")", n, a, b+1);
            }
        }
        
    }
    vector<string> generateParenthesis(int n) {
        solve("", n, 0, 0);
        
        return res;
    }
};

File: LeetCode/medium/Generate_Parentheses_recursive.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        
        def solve(cur, no, l):
            if len(cur) == n*2:
                res.append(cur)
            else:
                if l == 0:
                    if no > 0:
                        solve(cur+"(", no-1, l+1)
                elif l == n:
                    solve(cur+")", no, l-1)
                else:
                    if no > 0:
                        solve(cur+"(", no-1, l+1)
                    solve(cur+")", no, l-1)
        res = []
        solve("", n, 0)
        return res

